>  runbot seems a bit contaminated.
It is. I had to uninstall sale and install the modules again.

When I try to open the wizard I get this exception:

```python
Odoo Server Error
Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 98, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 962, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 759, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 746, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/models.py", line 5533, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/.repo_requirements/odoo/odoo/models.py", line 5377, in _onchange_eval
    method_res = method(self)
  File "/home/odoo/build/OCA/sale-workflow/sale_order_product_recommendation/wizards/sale_order_recommendation.py", line 108, in _generate_recommendations
    self._prepare_recommendation_line_vals(found_line, line)
  File "/home/odoo/build/OCA/sale-workflow/sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py", line 15, in _prepare_recommendation_line_vals
    group_line, so_line=so_line)
  File "/home/odoo/build/OCA/sale-workflow/sale_order_product_recommendation/wizards/sale_order_recommendation.py", line 67, in _prepare_recommendation_line_vals
    "product_id": group_line["product_id"][0],
KeyError: 'product_id'
```

--- a/sale_order_product_recommendation/README.rst
+++ b/sale_order_product_recommendation/README.rst
@@ -76,6 +76,7 @@ Contributors
 
   * Jairo Llopis <jairo.llopis@tecnativa.com>
   * David Vidal <david.vidal@tecnativa.com>
+  * Pedro M. Baeza
 
 Maintainers
 ~~~~~~~~~~~

--- a/sale_order_product_recommendation/__manifest__.py
+++ b/sale_order_product_recommendation/__manifest__.py
@@ -1,4 +1,5 @@
 # Copyright 2017 Jairo Llopis <jairo.llopis@tecnativa.com>
+# Copyright 2020 Tecnativa - Pedro M. Baeza
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 {
     "name": "Sale Order Product Recommendation",

--- a/sale_order_product_recommendation/readme/CONTRIBUTORS.rst
+++ b/sale_order_product_recommendation/readme/CONTRIBUTORS.rst
@@ -2,3 +2,4 @@
 
   * Jairo Llopis <jairo.llopis@tecnativa.com>
   * David Vidal <david.vidal@tecnativa.com>
+  * Pedro M. Baeza

--- a/sale_order_product_recommendation/static/description/index.html
+++ b/sale_order_product_recommendation/static/description/index.html
@@ -421,6 +421,7 @@ If you spotted it first, help us smashing it by providing a detailed and welcome
 <li><a class="reference external" href="https://www.tecnativa.com">Tecnativa</a>:<ul>
 <li>Jairo Llopis &lt;<a class="reference external" href="mailto:jairo.llopis&#64;tecnativa.com">jairo.llopis&#64;tecnativa.com</a>&gt;</li>
 <li>David Vidal &lt;<a class="reference external" href="mailto:david.vidal&#64;tecnativa.com">david.vidal&#64;tecnativa.com</a>&gt;</li>
+<li>Pedro M. Baeza</li>
 </ul>
 </li>
 </ul>

--- a/sale_order_product_recommendation/tests/test_recommendation.py
+++ b/sale_order_product_recommendation/tests/test_recommendation.py
@@ -1,4 +1,5 @@
 # Copyright 2017 Tecnativa - Jairo Llopis
+# Copyright 2020 Tecnativa - Pedro M. Baeza
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 from .test_recommendation_common import RecommendationCase
 from odoo.exceptions import UserError
@@ -58,7 +59,6 @@ class RecommendationCaseTests(RecommendationCase):
         wiz_line_prod1 = wizard.line_ids.filtered(
             lambda x: x.product_id == self.prod_1)
         wiz_line_prod1.units_included = qty
-        wiz_line_prod1._onchange_units_included()
         wizard.action_accept()
         self.assertEqual(len(self.new_so.order_line), 1)
         self.assertEqual(self.new_so.order_line.product_id, self.prod_1)
@@ -69,7 +69,6 @@ class RecommendationCaseTests(RecommendationCase):
         wiz_line = wizard.line_ids.filtered(
             lambda x: x.product_id == self.prod_1)
         wiz_line.units_included = 0
-        wiz_line._onchange_units_included()
         # The confirmed line can't be deleted
         with self.assertRaises(UserError):
             wizard.action_accept()
@@ -86,7 +85,6 @@ class RecommendationCaseTests(RecommendationCase):
         wiz_line = wizard.line_ids.filtered(
             lambda x: x.product_id == self.prod_1)
         wiz_line.units_included = qty + 2
-        wiz_line._onchange_units_included()
         wizard.action_accept()
         # Deliver extra qty and make a new invoice
         self.new_so.order_line.qty_delivered = qty + 2

--- a/sale_order_product_recommendation/tests/test_recommendation_common.py
+++ b/sale_order_product_recommendation/tests/test_recommendation_common.py
@@ -7,6 +7,8 @@ class RecommendationCase(SavepointCase):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
+        # Make sure user has UoM activated for Forms to work
+        cls.env.user.groups_id = [(4, cls.env.ref('uom.group_uom').id)]
         cls.partner = cls.env['res.partner'].create({
             'name': 'Mr. Odoo',
         })

--- a/sale_order_product_recommendation/wizards/sale_order_recommendation.py
+++ b/sale_order_product_recommendation/wizards/sale_order_recommendation.py
@@ -179,6 +179,7 @@ class SaleOrderRecommendationLine(models.TransientModel):
     sale_line_id = fields.Many2one(
         comodel_name="sale.order.line",
     )
+    sale_uom_id = fields.Many2one(related="sale_line_id.product_uom")
     is_modified = fields.Boolean()
 
     @api.depends("partner_id", "product_id", "pricelist_id", "units_included")

--- a/sale_order_product_recommendation/wizards/sale_order_recommendation.py
+++ b/sale_order_product_recommendation/wizards/sale_order_recommendation.py
@@ -5,6 +5,7 @@
 
 from datetime import datetime, timedelta
 from odoo import api, fields, models
+from odoo.tests import Form
 
 
 class SaleOrderRecommendation(models.TransientModel):
@@ -128,28 +129,29 @@ class SaleOrderRecommendation(models.TransientModel):
 
     def action_accept(self):
         """Propagate recommendations to sale order."""
-        so_line_obj = self.env["sale.order.line"]
-        so_line_ids = []
         sequence = max(self.order_id.mapped('order_line.sequence') or [0])
+        order_form = Form(self.order_id)
+        to_remove = []
         for wiz_line in self.line_ids.filtered(
             lambda x: x.sale_line_id or x.units_included
         ):
-            # Use preexisting line if any
             if wiz_line.sale_line_id:
+                index = self.order_id.order_line.ids.index(
+                    wiz_line.sale_line_id.id)
                 if wiz_line.units_included:
-                    wiz_line.sale_line_id.update(
-                        wiz_line._prepare_update_so_line())
-                    wiz_line.sale_line_id.product_uom_change()
+                    with order_form.order_line.edit(index) as line_form:
+                        wiz_line._prepare_update_so_line(line_form)
                 else:
-                    wiz_line.sale_line_id.unlink()
-                continue
-            sequence += 1
-            # Use a new in-memory line otherwise
-            so_line = so_line_obj.new(
-                wiz_line._prepare_new_so_line(sequence))
-            so_line = wiz_line._trigger_so_line_onchanges(so_line)
-            so_line_ids.append(so_line.id)
-        self.order_id.order_line += so_line_obj.browse(so_line_ids)
+                    to_remove.append(index)
+            else:
+                sequence += 1
+                with order_form.order_line.new() as line_form:
+                    wiz_line._prepare_new_so_line(line_form, sequence)
+        # Remove at the end and in reverse order for not having problems
+        to_remove.reverse()
+        for index in to_remove:
+            order_form.order_line.remove(index)
+        order_form.save()
 
 
 class SaleOrderRecommendationLine(models.TransientModel):
@@ -204,23 +206,12 @@ class SaleOrderRecommendationLine(models.TransientModel):
                 quantity=one.units_included,
             ).price
 
-    def _prepare_update_so_line(self):
-        """So we can extend PO update"""
-        return {
-            "product_uom_qty": self.units_included,
-        }
+    def _prepare_update_so_line(self, line_form):
+        """So we can extend SO update"""
+        line_form.product_uom_qty = self.units_included
 
-    def _prepare_new_so_line(self, sequence):
-        """So we can extend PO create"""
-        return {
-            "order_id": self.wizard_id.order_id.id,
-            "product_id": self.product_id.id,
-            "sequence": sequence,
-        }
-
-    def _trigger_so_line_onchanges(self, so_line):
-        """Extensible method for trigger needed onchanges of the so ling"""
-        so_line.product_id_change()
-        so_line.product_uom_qty = self.units_included
-        so_line.product_uom_change()
-        return so_line
+    def _prepare_new_so_line(self, line_form, sequence):
+        """So we can extend SO create"""
+        line_form.product_id = self.product_id
+        line_form.sequence = sequence
+        line_form.product_uom_qty = self.units_included

--- a/sale_order_product_recommendation/wizards/sale_order_recommendation.py
+++ b/sale_order_product_recommendation/wizards/sale_order_recommendation.py
@@ -119,7 +119,9 @@ class SaleOrderRecommendation(models.TransientModel):
         so_line_obj = self.env["sale.order.line"]
         so_line_ids = []
         sequence = max(self.order_id.mapped('order_line.sequence') or [0])
-        for wiz_line in self.line_ids.filtered('is_modified'):
+        for wiz_line in self.line_ids.filtered(
+            lambda x: x.sale_line_id or x.units_included
+        ):
             # Use preexisting line if any
             if wiz_line.sale_line_id:
                 if wiz_line.units_included:
@@ -180,7 +182,6 @@ class SaleOrderRecommendationLine(models.TransientModel):
         comodel_name="sale.order.line",
     )
     sale_uom_id = fields.Many2one(related="sale_line_id.product_uom")
-    is_modified = fields.Boolean()
 
     @api.depends("partner_id", "product_id", "pricelist_id", "units_included")
     def _compute_price_unit(self):
@@ -190,10 +191,6 @@ class SaleOrderRecommendationLine(models.TransientModel):
                 pricelist=one.pricelist_id.id,
                 quantity=one.units_included,
             ).price
-
-    @api.onchange("units_included")
-    def _onchange_units_included(self):
-        self.is_modified = bool(self.sale_line_id or self.units_included)
 
     def _prepare_update_so_line(self):
         """So we can extend PO update"""

--- a/sale_order_product_recommendation/wizards/sale_order_recommendation_view.xml
+++ b/sale_order_product_recommendation/wizards/sale_order_recommendation_view.xml
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!-- Copyright 2017 Jairo Llopis <jairo.llopis@tecnativa.com>
+     Copyright 2020 Tecnativa - Pedro M. Baeza
      License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
 
 <odoo>
@@ -17,7 +18,7 @@
                         <field name="line_amount" widget="numeric_step"/>
                     </group>
                     <group>
-                        <field name="line_ids" nolabel="1">
+                        <field name="line_ids" nolabel="1" mode="tree,kanban">
                             <tree create="0" delete="0" editable="top">
                                 <field name="currency_id" invisible="1"/>
                                 <field name="sale_line_id" invisible="1"/>
@@ -28,6 +29,40 @@
                                 <field name="units_delivered"/>
                                 <field name="units_included" widget="numeric_step"/>
                             </tree>
+                            <kanban class="o_kanban_mobile" quick_create="0" create="0" delete="0">
+                                <field name="id"/>
+                                <field name="product_id"/>
+                                <field name="is_modified"/>
+                                <field name="sale_line_id"/>
+                                <field name="price_unit"/>
+                                <field name="sale_uom_id"/>
+                                <field name="times_delivered"/>
+                                <field name="units_delivered"/>
+                                <field name="units_included"/>
+                                <field name="currency_id"/>
+                                <templates>
+                                    <t t-name="kanban-box">
+                                        <div class="oe_kanban_global_click">
+                                            <div class="o_kanban_image">
+                                                <img t-att-src="kanban_image('product.template', 'image_small', record.product_id.raw_value)" alt="Product"/>
+                                            </div>
+                                            <div class="oe_kanban_details">
+                                                <div><field name="product_id"/></div>
+                                                <div>Price: <field name="price_unit" widget="monetary" options="{'currency_field': 'currency_id', 'field_digits': True}" class="mt-0 mb-0"/></div>
+                                                <div>Delivered: <field name="times_delivered"/> times - <field name="units_delivered"/> <field name="sale_uom_id"/></div>
+                                                <div><strong><span>To be included: </span><field name="units_included"/><span> </span><field name="sale_uom_id"/></strong></div>
+                                            </div>
+                                        </div>
+                                    </t>
+                                </templates>
+                            </kanban>
+                            <form>
+                                <group>
+                                    <field name="product_id" readonly="1" force_save="1"/>
+                                    <field name="price_unit"/>
+                                    <field name="units_included" widget="numeric_step"/>
+                                </group>
+                            </form>
                         </field>
                     </group>
                 </sheet>

--- a/sale_order_product_recommendation/wizards/sale_order_recommendation_view.xml
+++ b/sale_order_product_recommendation/wizards/sale_order_recommendation_view.xml
@@ -22,7 +22,6 @@
                             <tree create="0" delete="0" editable="top">
                                 <field name="currency_id" invisible="1"/>
                                 <field name="sale_line_id" invisible="1"/>
-                                <field name="is_modified" invisible="1"/>
                                 <field name="product_id" readonly="1" force_save="1"/>
                                 <field name="price_unit"/>
                                 <field name="times_delivered"/>
@@ -32,7 +31,6 @@
                             <kanban class="o_kanban_mobile" quick_create="0" create="0" delete="0">
                                 <field name="id"/>
                                 <field name="product_id"/>
-                                <field name="is_modified"/>
                                 <field name="sale_line_id"/>
                                 <field name="price_unit"/>
                                 <field name="sale_uom_id"/>

--- a/sale_order_product_recommendation_secondary_unit/tests/test_recommendations.py
+++ b/sale_order_product_recommendation_secondary_unit/tests/test_recommendations.py
@@ -1,4 +1,5 @@
 # Copyright 2019 Tecnativa - David Vidal
+# Copyright 2020 Tecnativa - Pedro M. Baeza
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 from odoo.addons.sale_order_product_recommendation.tests import (
     test_recommendation_common)
@@ -68,11 +69,9 @@ class RecommendationCaseTests(test_recommendation_common.RecommendationCase):
             lambda x: x.product_id == self.prod_2)
         wl_prod1.secondary_uom_qty = 2  # 20 units
         wl_prod1._onchange_secondary_uom()
-        wl_prod1._onchange_units_included()
         wl_prod2.secondary_uom_id = self.secondary_unit_p2
         wl_prod2.secondary_uom_qty = 2  # 48 units
         wl_prod2._onchange_secondary_uom()
-        wl_prod2._onchange_units_included()
         wizard.action_accept()
         self.assertEqual(len(self.new_so.order_line), 2)
         sl_p1 = self.new_so.order_line.filtered(
@@ -91,7 +90,6 @@ class RecommendationCaseTests(test_recommendation_common.RecommendationCase):
             lambda x: x.product_id == self.prod_1)
         wl_prod1.secondary_uom_qty = 1  # 10 units
         wl_prod1._onchange_secondary_uom()
-        wl_prod1._onchange_units_included()
         wizard.action_accept()
         sl_p1 = self.new_so.order_line.filtered(
             lambda x: x.product_id == self.prod_1)

--- a/sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py
+++ b/sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py
@@ -71,31 +71,16 @@ class SaleOrderRecommendationLine(models.TransientModel):
                 precision_rounding=self.secondary_uom_id.uom_id.rounding) != 0:
             self.secondary_uom_qty = qty
 
-    def _prepare_update_so_line(self):
-        res = super()._prepare_update_so_line()
-        if self.secondary_uom_id and self.secondary_uom_qty:
-            res.update({
-                'secondary_uom_id': self.secondary_uom_id.id,
-                'secondary_uom_qty': self.secondary_uom_qty,
-            })
+    def _prepare_update_so_line(self, line_form):
+        res = super()._prepare_update_so_line(line_form)
+        if self.secondary_uom_id:
+            line_form.secondary_uom_id = self.secondary_uom_id
+            line_form.secondary_uom_qty = self.secondary_uom_qty
         return res
 
-    def _prepare_new_so_line(self, sequence):
-        res = super()._prepare_new_so_line(sequence)
-        if self.secondary_uom_id and self.secondary_uom_qty:
-            res.update({
-                'secondary_uom_id': self.secondary_uom_id.id,
-                'secondary_uom_qty': self.secondary_uom_qty,
-            })
+    def _prepare_new_so_line(self, line_form, sequence):
+        res = super()._prepare_new_so_line(line_form, sequence)
+        if self.secondary_uom_id:
+            line_form.secondary_uom_id = self.secondary_uom_id
+            line_form.secondary_uom_qty = self.secondary_uom_qty
         return res
-
-    def _trigger_so_line_onchanges(self, so_line):
-        """We need to recompute secondary uom qty from the units in the line"""
-        secondary_uom_id = self.secondary_uom_id
-        so_line = super()._trigger_so_line_onchanges(so_line)
-        if secondary_uom_id and self.secondary_uom_qty:
-            # Prevent product defaults
-            if so_line.secondary_uom_id != secondary_uom_id:
-                so_line.secondary_uom_id = secondary_uom_id
-            so_line.onchange_secondary_unit_product_uom_qty()
-        return so_line

--- a/sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py
+++ b/sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py
@@ -13,6 +13,9 @@ class SaleOrderRecommendationLine(models.TransientModel):
         comodel_name='product.secondary.unit',
         related='product_id.sale_secondary_uom_id',
     )
+    secondary_uom_name = fields.Char(
+        related='secondary_uom_id.name',
+    )
     secondary_uom_qty = fields.Float(
         string='Secondary Qty',
         digits=dp.get_precision('Product Unit of Measure'),

--- a/sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation_view.xml
+++ b/sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation_view.xml
@@ -6,11 +6,25 @@
         <field name="model">sale.order.recommendation</field>
         <field name="inherit_id" ref="sale_order_product_recommendation.sale_order_recommendation_view_form"/>
         <field name="arch" type="xml">
-            <field name="units_included" position="before">
+            <xpath expr="//field[@name='line_ids']/tree//field[@name='units_included']" position="before">
                 <field name="product_tmpl_id" invisible="1"/>
                 <field name="secondary_uom_id" domain="[('product_tmpl_id', '=', product_tmpl_id)]"/>
                 <field name="secondary_uom_qty" attrs="{'invisible': [('secondary_uom_id', '=', False)]}" widget="numeric_step"/>
-            </field>
+            </xpath>
+            <xpath expr="//field[@name='line_ids']/kanban//field[@name='units_included']" position="before">
+                <field name="secondary_uom_qty"/>
+                <field name="secondary_uom_id"/>
+                <field name="secondary_uom_name"/>
+            </xpath>
+            <xpath expr="//field[@name='line_ids']/kanban//t[@t-name='kanban-box']//field[@name='units_included']" position="before">
+                <t t-if="record.secondary_uom_id.value">
+                    <field name="secondary_uom_qty"/><span> </span><field name="secondary_uom_name"/><span> - </span>
+                </t>
+            </xpath>
+            <xpath expr="//field[@name='line_ids']/form//field[@name='units_included']" position="before">
+                <field name="secondary_uom_id" attrs="{'invisible': [('secondary_uom_id', '=', False)]}" />
+                <field name="secondary_uom_qty" attrs="{'invisible': [('secondary_uom_id', '=', False)]}" widget="numeric_step"/>
+            </xpath>
         </field>
     </record>
 </odoo>

